### PR TITLE
Extreme points

### PR DESCRIPTION
**Pull request by kkatrio**
## Summary of Changes

convex_hull_3 overload which voids the construction of a polyhedron.
* Affected package(s):  convex_hull_3
* Small Feature:

### CODE REVIEW

File: Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h [hunk 1]
```diff
@@ -65,4 +65,35 @@ void convex_hull_3(InputIterator first, InputIterator last,
...
+\brief computes the convex hull of the set of points in the range.
+The result is put in an output iterator as a list of points.
+
+\tparam InputRange a range of `Traits::Point_3`, model of `Range`.
+\tparam OutputIterator must be an output iterator with a value type  equivalent to `Traits::Point_3`.
+\tparam Traits must be model of the concept `ConvexHullTraits_3`.
+For the purposes of checking the postcondition that the convex hull
+is valid, `Traits` must also be a model of the concept
+`IsStronglyConvexTraits_3`.   Furthermore, `Traits` must define a type `Polygon_mesh` that is a model of
+`MutableFaceGraph`.
```
Comment (sloriot) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h]: Do we really need `IsStronglyConvexTraits_3` and `Polygon_mesh` here?

File: Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h [hunk 2]
```diff
@@ -65,4 +65,35 @@ void convex_hull_3(InputIterator first, InputIterator last,
 Object& ch_object,
 const Traits& ch_traits = Default_traits);
 
+
+/*!
+\ingroup PkgConvexHull3Functions
+
+\brief computes the convex hull of the set of points in the range.
+The result is put in an output iterator as a list of points.
```
Comment (sloriot) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h]: I'd rephrase as something like:  put in `out` the point from `range` that are on its convex hull.

File: Convex_hull_3/test/Convex_hull_3/test_extreme_points.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  typedef Polyhedron_3::Point_iterator Point_iterator;
+  int i = 0;
+  for (Point_iterator p_it = polyhedron.points_begin(); p_it != polyhedron.points_end(); ++p_it)
+  {
+    Point_3 p = *p_it;
+    CGAL_assertion(p == extreme_points[i]);
+    i++;
+  }
+}
```
Comment (sloriot) [inline Convex_hull_3/test/Convex_hull_3/test_extreme_points.cpp]: Please also test we have the expected result for 1 point, 2 points, 3 collinear points, ...
I suspect that for 3 collinear points you'll have only 2 while we document 3 in `extreme_point_3`

File: Convex_hull_3/include/CGAL/convex_hull_3.h [hunk 4]
```diff
@@ -624,8 +689,43 @@ void non_coplanar_quickhull_3(std::list<typename Traits::Point_3>& points,
...
+  BOOST_FOREACH(Point_3 p, points)
+  {
+    vertex_descriptor v = add_vertex(P);
+    put(vpm, v, p);
+  }
+}
+
+template <class Point_3, class OutputIterator>
+void
+ch_triangle_3(Point_3& p1, Point_3& p2, Point_3& p3,
```
Comment (sloriot) [inline Convex_hull_3/include/CGAL/convex_hull_3.h]: no, add an overload for make_triangle close to the one added for make_tetrahedron

File: Convex_hull_3/include/CGAL/convex_hull_3.h [hunk 5]
```diff
@@ -857,45 +955,67 @@ template <class InputIterator, class Polyhedron_3, class Traits>
...
-  CGAL_ch_precondition(points.size() > 3);
+  clear(polyhedron);
 
-  // at least 4 points 
-  typename Traits::Collinear_3 collinear = traits.collinear_3_object();
-  typename Traits::Equal_3 equal = traits.equal_3_object();
+  if(points.size() == 0)
+    return;
+  else if(points.size() == 1 || points.size() == 2)
+    internal::Convex_hull_3::add_isolated_points(points, polyhedron);
```
Comment (sloriot) [inline Convex_hull_3/include/CGAL/convex_hull_3.h]: what if the two points are identical?

Comment (sloriot) [reply Convex_hull_3/include/CGAL/convex_hull_3.h]: actually I do not see why you changed the logic of the initial code. what if I have 100 identical points?

File: Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h [hunk 6]
```diff
@@ -8,27 +8,27 @@ namespace CGAL {
 the convex hull is stored in `pm`. Note that the convex hull will be triangulated,
 that is `pm` will contain only triangular facets.
 
-\pre There are at least four points in the range 
+\pre There are at least four points in the range
```
Comment (sloriot) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h]: Is this still true?

File: Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h [hunk 7]
```diff
@@ -65,4 +65,31 @@ void convex_hull_3(InputIterator first, InputIterator last,
 Object& ch_object,
 const Traits& ch_traits = Default_traits);
 
+
+/*!
+\ingroup PkgConvexHull3Functions
+
+\brief computes the convex hull of the set of points in the range.
```
Comment (sloriot) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h]: copy in `out` the points on the convex hull of the points in `range`.

File: Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h [hunk 8]
```diff
@@ -65,4 +65,31 @@ void convex_hull_3(InputIterator first, InputIterator last,
...
+It puts in out the points from `range` that are on its convex hull.
+
+\tparam InputRange a range of `Traits::Point_3`, model of `Range`.
```
Comment (sloriot) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h]: Range::const_iterator is model of what? does the range need to be mutable or ConstRange is fine?

File: Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h [hunk 9]
```diff
@@ -65,4 +65,31 @@ void convex_hull_3(InputIterator first, InputIterator last,
...
+\tparam OutputIterator must be an output iterator with a value type  equivalent to `Traits::Point_3`.
+\tparam Traits must be model of the concept `ConvexHullTraits_3`.
+
+If the kernel `R` of the points determined by the value type of `InputIterator`
+is a kernel with exact predicates but inexact constructions
+(in practice we check `R::Has_filtered_predicates_tag` is `Tag_true` and `R::FT` is a floating point type),
+then the default traits class of `convex_hull_3()` is `Convex_hull_traits_3<R>`, and `R` otherwise.
+
+@param range the range of input points.
+@param out an output iterator where the list of points will be put.
```
Comment (sloriot) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h]: list -> extreme

File: Convex_hull_3/include/CGAL/convex_hull_3.h [hunk 10]
```diff
@@ -63,9 +64,114 @@
...
+template <class Point_3, class OutputIterator>
+void add_isolated_points(const std::list<Point_3>& points,
+                         internal::Convex_hull_3::Output_iterator_wrapper<OutputIterator>& w)
+{
+  BOOST_FOREACH(Point_3 p, points)
+    ++w.out = p;
+}
+
+template <class Point_3, class Polyhedron_3>
+void add_isolated_points(const std::list<Point_3>& points, Polyhedron_3& P)
```
Comment (sloriot) [inline Convex_hull_3/include/CGAL/convex_hull_3.h]: I'd prefer this overload to be close to the function copy ch2 for PolygonMesh

File: Convex_hull_3/include/CGAL/convex_hull_3.h [hunk 11]
```diff
@@ -857,45 +963,78 @@ template <class InputIterator, class Polyhedron_3, class Traits>
...
-        "All points are equal; cannot construct polyhedron.");
-  
+  // if all points are equal, return empty set.
+  if(point2_it == points.end())
+    return;
+
+  if(points.size() == 2)
+  {
+    add_isolated_points(points, polyhedron);
+    return;
```
Comment (sloriot) [inline Convex_hull_3/include/CGAL/convex_hull_3.h]: this is useless and can be catch below

File: Convex_hull_3/include/CGAL/convex_hull_3.h [hunk 12]
```diff
@@ -857,45 +963,78 @@ template <class InputIterator, class Polyhedron_3, class Traits>
...
 
   Point_3_list points(first, beyond);
-  CGAL_ch_precondition(points.size() > 3);
+  clear(polyhedron);
+
+  if(points.size() == 1)
+  {
+    add_isolated_points(points, polyhedron);
+    return;
```
Comment (sloriot) [inline Convex_hull_3/include/CGAL/convex_hull_3.h]: useless as can be catch below

File: Convex_hull_3/include/CGAL/convex_hull_3.h [hunk 13]
```diff
@@ -857,45 +963,78 @@ template <class InputIterator, class Polyhedron_3, class Traits>
...
+  }
+
+  // if there are 3 points not equal, then the third is not at points.end()
+  if(points.size() == 3)
+  {
+    if(collinear(*point1_it,*point2_it,*point3_it))
+    {
+      std::list<Point_3> endpoints;
+      find_endpoints(points, endpoints, traits);
+      add_isolated_points(endpoints, polyhedron);
```
Comment (sloriot) [inline Convex_hull_3/include/CGAL/convex_hull_3.h]: here simply take the 3 points of the 3 iterators and the extreme one are the one that are le larget and the smaller one using a lexicographic sort (something like 2 if and swap should make the job)

**Comment by sloriot** [comment]
CHANGES.md need to be updated.

**Comment by lrineau** [comment]
@MaelRL Why did you re-add the "not yet approved" label?

**Comment by sloriot** [comment]
@maxGimeno noticed that with `cross.off` some extra points are kept on the convex hull (the number of points of the convex hull is not minimal). @afabri is it expected or is it a bug? If it is expected we need to indicate it in the documentation of `convex_hull_3()` overloads as well as this PR's functions.

**Comment by sloriot** [comment]
I should have mentioned that `CGAL::is_strongly_convex_3()` returns `false`.

File: Convex_hull_3/test/Convex_hull_3/test_extreme_points.cpp [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+  typedef typename boost::graph_traits<SurfaceMesh>::vertex_descriptor vertex_descriptor;
+  BOOST_FOREACH(vertex_descriptor v, vertices(mesh))
+  {
+    Point_3 p = get(vpmap, v);
+    mesh_points.push_back(p);
+  }
+
+  std::vector<Point_3> extreme_points;
+  CGAL::extreme_points_3(mesh_points, std::back_inserter(extreme_points)) ;
+  CGAL_assertion(extreme_points.size() == 8);
```
Comment (sloriot) [inline Convex_hull_3/test/Convex_hull_3/test_extreme_points.cpp]: This seems to be luck, I've tried with only the points and got the same issue as with `cross.off`, not strictly convex.

**Comment by afabri** [comment]
Concerning the question feature or bug, points that are coplanar and sit inside the 2D convex hull of these points should not be in the 3D convex hull. Similar for collinear points.

**Comment by lrineau** [comment]
Note that there are two Travis errors:
- a compilation error: https://travis-ci.org/CGAL/cgal/jobs/395400333#L2287
- a license check error: https://travis-ci.org/CGAL/cgal/jobs/395400321#L1848

This PR has never been tested in `integration`, and does not seem ready to be tested. And the code freeze is in four days. Will it be postponed to CGAL 4.14?

**Comment by maxGimeno** [comment]
I strongly doubt it will be ready in 4 days, I'd postpone it.

**Comment by maxGimeno** [comment]
@sloriot I think your modification to the logic for the filtered predicates in Is_on_positive_side_3 broke the Protector type or something, leading to the static_assertion failure in the  quick_hull_default_traits test. Can you please look into it ?

File: Convex_hull_3/test/Convex_hull_3/quick_hull_default_traits.cpp [hunk 15]
```diff
@@ -26,6 +26,6 @@ int main()
   CGAL_static_assertion( (boost::is_same<SCR,Default_traits_for_Chull_3<SCR::Point_3>::type>::value) );
   CGAL_static_assertion( (boost::is_same<EPEC,Default_traits_for_Chull_3<EPEC::Point_3>::type>::value) );
   CGAL_static_assertion( (boost::is_same<CGAL::Convex_hull_traits_3<EPIC, CGAL::Default, CGAL::Tag_true>,Default_traits_for_Chull_3<EPIC::Point_3>::type>::value) );
-  CGAL_static_assertion( (boost::is_same<Is_on_positive_side_of_plane_3<CGAL::Convex_hull_traits_3<EPIC, CGAL::Default, CGAL::Tag_true> >::Protector,CGAL::Protect_FPU_rounding<true> >::value) );
+  CGAL_static_assertion( (boost::is_same<Is_on_positive_side_of_plane_3<CGAL::Convex_hull_traits_3<EPIC, CGAL::Default, CGAL::Tag_true>, boost::true_type >::Protector,CGAL::Protect_FPU_rounding<true> >::value) );
```
Comment (lrineau) [inline Convex_hull_3/test/Convex_hull_3/quick_hull_default_traits.cpp]: @sloriot, can you review this patch please?

**Comment by sloriot** [comment]
Note that the bug in Convex hull about extra points on the convex hull is not related to this PR so it is not a blocker.

File: Convex_hull_3/test/Convex_hull_3/quick_hull_default_traits.cpp (hunk 15 above)
Comment (sloriot) [reply Convex_hull_3/test/Convex_hull_3/quick_hull_default_traits.cpp]: Looks good!

**Comment by sloriot** [comment]
`convex_hull_traits_3_fp_bug.cpp` is failling. I've added an extra assertion that is showing that the output is not strictly convex.

**Comment by sloriot** [comment]
@maxGimeno, I think this shows that `6e39f1aa056bdf4e2e6c3ad116fb56c98e46f5da` is not sufficient to fix the issue. I suggest that you revert that patch and that we leave this PR as is and that we fill an issue to fix the bug of the strict convexity of  `Convex_hull_3`. Actually I even think that my last commit should be undone since I think 82 is not the minimal number of points on the convex hull in the test mentioned above.

**Comment by maxGimeno** [comment]
Done :(

File: Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h [hunk 16]
```diff
@@ -65,4 +63,66 @@ void convex_hull_3(InputIterator first, InputIterator last,
...
+If the kernel `R` of the points from `InputRange`
+is a kernel with exact predicates but inexact constructions
+(in practice we check `R::Has_filtered_predicates_tag` is `Tag_true` and `R::FT` is a floating point type),
+then the default traits class used is `Convex_hull_traits_3<R>`, and `R` otherwise.
+
+@param range the range of input points.
+@param out an output iterator where the extreme points will be put.
+@param traits an instance of `Traits`.
+
+\sa CGAL::Extreme_points_traits_adapter_3
```
Comment (sloriot) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h]: missing backticks

File: Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h [hunk 17]
```diff
@@ -65,4 +63,66 @@ void convex_hull_3(InputIterator first, InputIterator last,
...
+@param map the map containing the Vertices and their associated points.
+@param out an output iterator where the extreme vertices will be put.
+@param traits an instance of `Traits`.
+*/
+template <class InputRange,
+          class OutputIterator,
+          class PointPropertyMap,
+          class Traits>
+OutputIterator
+extreme_vertices(const InputRange& range,
```
Comment (sloriot) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h]: @maxGimeno  I think this function should be gone now?

File: Convex_hull_3/include/CGAL/Extreme_points_traits_adapter_3.h [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+    return static_cast<const F*>(this)->operator()(
+          get(vpm_,p), 
+          get(vpm_,q),
+          get(vpm_,r), 
+          get(vpm_,s));
+  }
+};
+}//end Convex_hull_impl
+template<class Base_traits,class PointPropertyMap>
+class Extreme_points_traits_adapter_3
```
Comment (sloriot) [inline Convex_hull_3/include/CGAL/Extreme_points_traits_adapter_3.h]: I would put the property map as first parameter and Base_traits default could be extracted from `internal::Convex_hull_3::Default_traits_for_Chull_3` with `boost::property_traits<PointPropertyMap>::reference` as `Point_3`. Documentation wise, you can copy-paste the doc from `convex_hull_3`: _If the kernel `R` of the points determined  ...._

Comment (sloriot) [reply Convex_hull_3/include/CGAL/Extreme_points_traits_adapter_3.h]: A helper function `make_extreme_points_traits_adapter` would be nice to have too.

File: Installation/CHANGES.md [hunk 19]
```diff
@@ -49,6 +49,11 @@ Release date: September 2018
     to reflect the real needs of the code (some types and operators were used
     in the code but did not appear in the concepts).
 
+### Convex hull 3
+
+-   Added the function `extreme_points_3()` computing the
+    points on the convex hull without underlying connectivity.
```
Comment (sloriot) [inline Installation/CHANGES.md]: Maybe mention the traits here and what it enables.

**Comment by sloriot** [comment]
An example would be nice to have + mention in the user manual.

**Comment by lrineau** [comment]
@sloriot You seems in charge of the PR, now. I have two questions:

- Is that a small feature that should be in CGAL-4.13?
- Is it really reader to be tested? You added the label "Ready to be tested" one month ago, but important changes and reviews were done since then. Is that finished?

**Comment by sloriot** [comment]
I think the PR is ready now. I have no reason to have it in 4.13 except that I think it is a nice addition regularly asked by users. No pb for me to have in in 4.14.

**Comment by sloriot** [comment]
The two commits above should fix the error and warnings in the testsuite.

**Comment by sloriot** [comment]
CGAL-4.13-Ic-57 tests look good, waiting for the 15 days pre-approval period to be over.

**Comment by lrineau** [comment]
After the merge of #3226 in `master`, this PR is now in conflict.

**Comment by lrineau** [comment]
Successfully tested in [CGAL-4.13-Ic-68][*], and then merged.

  [*]: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.13-Ic-68.shtml
